File: sdks/python/stepflow-py/src/stepflow_py/worker/blob_ref.py
# Copyright 2025 DataStax Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License"); you may not
# use this file except in compliance with the License. You may obtain a copy of
# the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
# License for the specific language governing permissions and limitations under
# the License.

"""Blob reference (`$blob` sentinel) for representing blobified data in JSON values.

A blob ref is a JSON object with a ``$blob`` key that indicates "this value is stored
in the blob store". This is a runtime data convention, distinct from workflow-definition
ValueExpr (``$step``, ``$input``, ``$variable``).

Example::

    {"$blob": "<sha256hex>", "blobType": "data", "size": 12345}
"""

from __future__ import annotations

import json
import logging
from dataclasses import dataclass
from typing import Any

from stepflow_py.api.models.blob_type import BlobType

logger = logging.getLogger(__name__)

BLOB_REF_KEY = "$blob"


@dataclass
class BlobRef:
    """A reference to a blob stored in the blob store."""

    blob_id: str
    blob_type: BlobType = BlobType.DATA
    size: int | None = None

    def to_dict(self) -> dict[str, Any]:
        """Convert to a JSON-serializable dict."""
        d: dict[str, Any] = {BLOB_REF_KEY: self.blob_id}
        if self.blob_type != BlobType.DATA:
            d["blobType"] = self.blob_type.value
        if self.size is not None:
            d["size"] = self.size
        return d

    @classmethod
    def from_dict(cls, d: dict[str, Any]) -> BlobRef | None:
        """Try to parse a dict as a blob ref. Returns None if not a blob ref."""
        if not isinstance(d, dict) or BLOB_REF_KEY not in d:
            return None
        blob_id = d[BLOB_REF_KEY]
        if not isinstance(blob_id, str) or len(blob_id) != 64:
            return None
        blob_type_str = d.get("blobType", "data")
        try:
            blob_type = BlobType(blob_type_str)
        except ValueError:
            blob_type = BlobType.DATA
        size = d.get("size")
        return cls(blob_id=blob_id, blob_type=blob_type, size=size)


def is_blob_ref(value: Any) -> bool:
    """Quick check whether a value is a blob ref (dict with ``$blob`` key)."""
    return isinstance(value, dict) and BLOB_REF_KEY in value


async def blobify_inputs(
    input_data: Any,
    threshold: int,
) -> tuple[Any, list[str]]:
    """Replace large top-level fields in an input dict with blob refs.

    Checks each top-level field individually against the threshold.
    Fields whose JSON serialization exceeds the threshold are stored as blobs
    and replaced with blob refs.

    Args:
        input_data: The input dict (or any value — non-dicts pass through).
        threshold: Byte size threshold. 0 means disabled.

    Returns:
        A tuple of (possibly modified input, list of created blob IDs).
    """
    from stepflow_py.worker import blob_store

    if threshold == 0 or not isinstance(input_data, dict):
        return input_data, []

    created_ids: list[str] = []
    result: dict[str, Any] = {}

    for key, value in input_data.items():
        if is_blob_ref(value):
            result[key] = value
            continue
        serialized = json.dumps(value, separators=(",", ":")).encode("utf-8")
        size = len(serialized)
        if size > threshold:
            blob_id = await blob_store.put_blob(value)
            ref = BlobRef(blob_id=blob_id, blob_type=BlobType.DATA, size=size)
            result[key] = ref.to_dict()
            created_ids.append(blob_id)
            logger.debug("Blobified field %r (%d bytes) -> %s", key, size, blob_id[:12])
        else:
            result[key] = value

    return result, created_ids


def _collect_blob_ids(value: Any, ids: set[str]) -> None:
    """Walk a JSON value tree and collect all blob ref IDs."""
    if isinstance(value, dict):
        ref = BlobRef.from_dict(value)
        if ref is not None:
            ids.add(ref.blob_id)
        else:
            for v in value.values():
                _collect_blob_ids(v, ids)
    elif isinstance(value, list):
        for item in value:
            _collect_blob_ids(item, ids)


def _replace_blob_refs(value: Any, resolved: dict[str, Any]) -> Any:
    """Replace blob refs with their pre-fetched data (no I/O)."""
    if isinstance(value, dict):
        ref = BlobRef.from_dict(value)
        if ref is not None and ref.blob_id in resolved:
            return resolved[ref.blob_id]
        return {k: _replace_blob_refs(v, resolved) for k, v in value.items()}
    if isinstance(value, list):
        return [_replace_blob_refs(item, resolved) for item in value]
    return value


async def resolve_blob_refs(
    value: Any,
    _depth: int = 0,
) -> Any:
    """Replace blob refs in a JSON value with their inline data.

    All blob refs at the current level are fetched in parallel using
    :func:`~stepflow_py.worker.blob_store.get_blobs`, then substituted.
    If the resolved data itself contains blob refs, another round of
    parallel fetching is performed (up to ``max_depth`` rounds).

    Args:
        value: The JSON value to resolve.

    Returns:
        The value with all blob refs replaced by their inline data.
    """
    from stepflow_py.worker import blob_store

    max_depth = 8
    if _depth >= max_depth:
        return value

    # Collect all blob IDs in the current tree
    ids: set[str] = set()
    _collect_blob_ids(value, ids)
    if not ids:
        return value

    logger.debug("Resolving %d blob ref(s) in parallel", len(ids))

    # Batch-fetch all blobs in parallel
    resolved = await blob_store.get_blobs(ids)

    # Replace refs with fetched data
    result = _replace_blob_refs(value, resolved)

    # Resolved data may itself contain blob refs — check for another round
    next_ids: set[str] = set()
    _collect_blob_ids(result, next_ids)
    if next_ids:
        return await resolve_blob_refs(result, _depth + 1)
    return result

File: sdks/rust/crates/stepflow-worker/src/blob_ref.rs
// Copyright 2025 DataStax Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
// in compliance with the License. You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software distributed under the License
// is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
// or implied. See the License for the specific language governing permissions and limitations under
// the License.

//! Blob reference (`$blob` sentinel) for auto-blobification of large values.
//!
//! A blob ref is a JSON object with a `$blob` key indicating the value is stored
//! in the blob store. This module provides:
//!
//! - **Input de-blobification**: `resolve_blob_refs` recursively walks a JSON tree,
//!   finds blob refs, fetches their data in parallel, and replaces them inline.
//! - **Output blobification**: `blobify_output` checks each top-level field of an
//!   output object and stores large fields as blobs, replacing them with blob refs.
//!
//! # Blob ref format
//!
//! ```json
//! {"$blob": "<64-char-sha256-hex>", "blobType": "data", "size": 12345}
//! ```

use serde_json::{Map, Value};
use tracing::debug;

use crate::ComponentContext;
use crate::error::ContextError;

/// The JSON key used to identify blob references.
const BLOB_REF_KEY: &str = "$blob";

/// Maximum recursion depth when resolving blob refs (prevents infinite loops
/// if resolved data itself contains blob refs).
const MAX_RESOLVE_DEPTH: usize = 8;

/// Check whether a JSON value is a blob ref (object with `$blob` key containing
/// a 64-character hex string).
fn is_blob_ref(value: &Value) -> Option<&str> {
    let obj = value.as_object()?;
    let blob_id = obj.get(BLOB_REF_KEY)?.as_str()?;
    if blob_id.len() == 64 {
        Some(blob_id)
    } else {
        None
    }
}

/// Collect all blob ref IDs from a JSON value tree.
fn collect_blob_ids(value: &Value, ids: &mut Vec<String>) {
    match value {
        Value::Object(obj) => {
            if let Some(blob_id) = is_blob_ref(&Value::Object(obj.clone())) {
                ids.push(blob_id.to_string());
            } else {
                for v in obj.values() {
                    collect_blob_ids(v, ids);
                }
            }
        }
        Value::Array(arr) => {
            for item in arr {
                collect_blob_ids(item, ids);
            }
        }
        _ => {}
    }
}

/// Replace blob refs in a JSON value with their pre-fetched data (no I/O).
fn replace_blob_refs(value: Value, resolved: &std::collections::HashMap<String, Value>) -> Value {
    match value {
        Value::Object(obj) => {
            // Check if this object itself is a blob ref
            if let Some(blob_id) = obj.get(BLOB_REF_KEY).and_then(|v| v.as_str())
                && blob_id.len() == 64
                && let Some(data) = resolved.get(blob_id)
            {
                return data.clone();
            }
            // Otherwise recurse into fields
            let new_obj: Map<String, Value> = obj
                .into_iter()
                .map(|(k, v)| (k, replace_blob_refs(v, resolved)))
                .collect();
            Value::Object(new_obj)
        }
        Value::Array(arr) => Value::Array(
            arr.into_iter()
                .map(|v| replace_blob_refs(v, resolved))
                .collect(),
        ),
        other => other,
    }
}

/// Resolve all blob refs in a JSON value by fetching their data from the blob store.
///
/// Blob refs at all levels are collected and fetched in parallel, then substituted.
/// If resolved data itself contains blob refs, another round of fetching is performed
/// (up to [`MAX_RESOLVE_DEPTH`] rounds).
pub(crate) async fn resolve_blob_refs(
    value: Value,
    ctx: &ComponentContext,
) -> Result<Value, ContextError> {
    resolve_blob_refs_inner(value, ctx, 0).await
}

fn resolve_blob_refs_inner<'a>(
    value: Value,
    ctx: &'a ComponentContext,
    depth: usize,
) -> futures::future::BoxFuture<'a, Result<Value, ContextError>> {
    Box::pin(async move {
        if depth >= MAX_RESOLVE_DEPTH {
            return Ok(value);
        }

        // Collect all blob IDs in the current tree
        let mut ids = Vec::new();
        collect_blob_ids(&value, &mut ids);
        ids.sort_unstable();
        ids.dedup();

        if ids.is_empty() {
            return Ok(value);
        }

        debug!(count = ids.len(), depth, "Resolving blob refs in parallel");

        // Fetch all blobs in parallel
        let fetches: Vec<_> = ids
            .iter()
            .map(|id| {
                let ctx = ctx.clone();
                let id = id.clone();
                async move {
                    let data = ctx.get_blob(&id).await?;
                    Ok::<_, ContextError>((id, data))
                }
            })
            .collect();

        let results = futures::future::try_join_all(fetches).await?;
        let resolved: std::collections::HashMap<String, Value> = results.into_iter().collect();

        // Replace refs with fetched data
        let result = replace_blob_refs(value, &resolved);

        // Check if resolved data contains more blob refs
        let mut next_ids = Vec::new();
        collect_blob_ids(&result, &mut next_ids);
        if !next_ids.is_empty() {
            return resolve_blob_refs_inner(result, ctx, depth + 1).await;
        }

        Ok(result)
    })
}

/// Replace large top-level fields in an output object with blob refs.
///
/// Each top-level field is individually checked against `threshold`. Fields whose
/// JSON serialization exceeds the threshold are stored as blobs and replaced with
/// blob refs. Non-object values pass through unchanged.
///
/// Returns the (possibly modified) output value.
pub(crate) async fn blobify_output(
    output: Value,
    threshold: usize,
    ctx: &ComponentContext,
) -> Result<Value, ContextError> {
    if threshold == 0 {
        return Ok(output);
    }

    let obj = match output {
        Value::Object(obj) => obj,
        other => return Ok(other),
    };

    let mut result = Map::with_capacity(obj.len());

    for (key, value) in obj {
        // Skip values that are already blob refs
        if is_blob_ref(&value).is_some() {
            result.insert(key, value);
            continue;
        }

        // Check serialized size of this field
        let serialized = serde_json::to_string(&value)?;
        let size = serialized.len();

        if size > threshold {
            let blob_id = ctx.put_blob(value).await?;
            debug!(field = %key, size, blob_id = %&blob_id[..12], "Blobified output field");

            let mut ref_obj = Map::new();
            ref_obj.insert(BLOB_REF_KEY.to_string(), Value::String(blob_id));
            ref_obj.insert("blobType".to_string(), Value::String("data".to_string()));
            ref_obj.insert("size".to_string(), Value::Number(size.into()));
            result.insert(key, Value::Object(ref_obj));
        } else {
            result.insert(key, value);
        }
    }

    Ok(Value::Object(result))
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    #[test]
    fn test_is_blob_ref_valid() {
        let hex = "a".repeat(64);
        let value = json!({"$blob": hex});
        assert_eq!(is_blob_ref(&value), Some(hex.as_str()));
    }

    #[test]
    fn test_is_blob_ref_wrong_length() {
        let value = json!({"$blob": "tooshort"});
        assert_eq!(is_blob_ref(&value), None);
    }

    #[test]
    fn test_is_blob_ref_missing_key() {
        let value = json!({"other": "value"});
        assert_eq!(is_blob_ref(&value), None);
    }

    #[test]
    fn test_is_blob_ref_not_object() {
        assert_eq!(is_blob_ref(&json!("string")), None);
        assert_eq!(is_blob_ref(&json!(42)), None);
        assert_eq!(is_blob_ref(&json!(null)), None);
    }

    #[test]
    fn test_collect_blob_ids_nested() {
        let hex1 = "a".repeat(64);
        let hex2 = "b".repeat(64);
        let value = json!({
            "field1": {"$blob": hex1},
            "field2": {
                "nested": {"$blob": hex2}
            },
            "field3": "plain"
        });
        let mut ids = Vec::new();
        collect_blob_ids(&value, &mut ids);
        ids.sort();
        assert_eq!(ids, vec![hex1, hex2]);
    }

    #[test]
    fn test_collect_blob_ids_in_array() {
        let hex = "c".repeat(64);
        let value = json!([{"$blob": hex}, "plain", 42]);
        let mut ids = Vec::new();
        collect_blob_ids(&value, &mut ids);
        assert_eq!(ids, vec![hex]);
    }

    #[test]
    fn test_collect_blob_ids_empty() {
        let value = json!({"field": "value", "nested": {"a": 1}});
        let mut ids = Vec::new();
        collect_blob_ids(&value, &mut ids);
        assert!(ids.is_empty());
    }

    #[test]
    fn test_replace_blob_refs() {
        let hex = "d".repeat(64);
        let value = json!({
            "field1": {"$blob": hex},
            "field2": "unchanged"
        });
        let mut resolved = std::collections::HashMap::new();
        resolved.insert(hex, json!({"resolved": "data"}));

        let result = replace_blob_refs(value, &resolved);
        assert_eq!(
            result,
            json!({
                "field1": {"resolved": "data"},
                "field2": "unchanged"
            })
        );
    }

    #[test]
    fn test_replace_blob_refs_unresolved() {
        let hex = "e".repeat(64);
        let value = json!({"$blob": hex});
        let resolved = std::collections::HashMap::new();
        // Unresolved blob refs pass through unchanged
        let result = replace_blob_refs(value.clone(), &resolved);
        assert_eq!(result, value);
    }

    #[test]
    fn test_replace_blob_refs_nested_in_array() {
        let hex = "f".repeat(64);
        let value = json!([{"$blob": hex}, "keep"]);
        let mut resolved = std::collections::HashMap::new();
        resolved.insert(hex, json!("fetched"));

        let result = replace_blob_refs(value, &resolved);
        assert_eq!(result, json!(["fetched", "keep"]));
    }
}
